.repo/sample-list-generator/src/api-detector.ts

```diff
@@ -0,0 +1,144 @@
+import { IApiItem, TApiTypeResult, TExtensionApiMap } from './types';
+import * as babel from '@babel/core';
+import fs from 'fs/promises';
+import { getAllFiles, loadExtensionApis, singularize } from './utils';
+
+let EXTENSION_API_MAP: TExtensionApiMap = {};
+(async () => {
+  EXTENSION_API_MAP = await loadExtensionApis();
+})();
+
+export const getApiListForSample = async (
+  folderPath: string
+): Promise<IApiItem[]> => {
+  // get all js files in the folder
+  const jsFiles = (await getAllFiles(folderPath)).filter((file) =>
+    file.endsWith('.js')
+  );
+
+  const apis: Record<string, Set<string>> = {};
+
+  for (let file of jsFiles) {
+    const apiCalls = await extractApiCalls(await fs.readFile(file));
+    Object.keys(apiCalls).forEach((apiType) => {
+      if (!apis[apiType]) {
+        apis[apiType] = new Set();
+      }
+      apiCalls[apiType].forEach((api) => {
+        apis[apiType].add(api);
+      });
+    });
+  }
+
+  const result: IApiItem[] = [];
+
+  Object.keys(apis).forEach((apiType) => {
+    apis[apiType].forEach((apiName) => {
+      result.push({
+        type: singularize(apiType) as TApiTypeResult,
+        catagory: apiName.split('.')[0],
+        name: apiName.split('.')[1]
+      });
+    });
+  });
+
+  return result;
+};
+
+export const extractApiCalls = (
+  file: Buffer
+): Promise<Record<string, string[]>> => {
+  return new Promise((resolve) => {
+    const calls: Record<string, string[]> = {};
+
+    babel.parse(
+      file.toString('utf8'),
+      { ast: true, compact: false },
+      (err, result) => {
+        if (err || !result) {
+          resolve(calls);
+          return;
+        }
+
+        babel.traverse(result, {
```

Comment:
This is quite a long method, can you please split it up to make it easier to understand.